runtime/pkg/singleflight/singleflight.go

```diff
@@ -0,0 +1,48 @@
+// Package singleflight provides a duplicate function call suppression
+// mechanism. Borrowed from golang/groupcache.
+package singleflight
```

Comment:
The `groupcache` has a limitation that key has to be a `string`. We are allowing `any` to be key.

Reply:
The key we use is also a `string`, so isn't it fine?

Reply:
Although we are setting key as `string` as of now but it can be any value subject to some constraints. We can change the `querycache.getOrLoad` key method to `string` as well if required. I am fine with both approaches. Please let me know what do you think.

Reply:
If we're only using it as a `string`, I think we should use that type. There would also be a slight performance advantage.

As for vendoring vs. importing, I'm fine with both approaches when it's so little code

Reply:
Just realised based on IDE auto import that the impl has also been added to `golang.org/x/sync/singleflight`pkg with some changes. We can use that itself. This makes the decision between vendoring and importing easy.